### ISSUE TITLE
False positive E1126 when index previously assigned to None

### ISSUE THREAD

**Issue by chrisjbillington** [PyCQA/pylint#2880]
The following code:
```python
N = 10
order = [0] * N
j = None
for i in range(N):
    if j is not None:
        order[j] += 1 # <--- E1126 for this line
    else:
        j = i
```

gives:

```
test.py:6:0: E1126: Sequence index is not an int, slice, or instance with __index__ (invalid-sequence-index)
```
I can see that this is because j was initially set to `None`, which would not be a valid index. However the code is actually correct, as a human can see, since it checks for `None` before using the index.

I'm not sure if there's anything pylint can do abut this, since I'm not sure how sophisticated it aims to be in checking this sort of thing. But reporting this anyway for what it's worth - feel free to close if this is outside the scope of pylint.

```$ pylint --version
pylint 2.2.2
astroid 2.1.0
Python 3.7.3 (default, Mar 26 2019, 21:43:19) 
[GCC 8.2.1 20181127]
```

**Comment by PCManticore** [comment]
Thanks for opening an issue. This is related to `pylint` not being able to understand control flow, a tricky topic for a non-human tool.

**Comment by mbyrnepr2** [comment]
This one no longer reproduces and should be fixed as per https://github.com/PyCQA/astroid/pull/1189.

Tested with:
pylint 2.17.0-dev0
astroid 2.15.0-dev0
Python 3.10.4
